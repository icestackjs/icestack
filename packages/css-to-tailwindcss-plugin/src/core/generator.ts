import * as t from '@babel/types'
import babelGenerate from '@babel/generator'
import type { Node, Rule } from 'postcss'
import type { GeneratorOptions } from '@babel/generator'
import { layerNodesKeys } from '../constants'
import type { LayerEnumType } from '../constants'
import type { BaseContext } from './base-context'
// addBase, addComponents, addUtilities, theme, addVariant, config, corePlugins, e, matchComponents, matchUtilities, matchVariant
// https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/setupContextUtils.js#L287
// resolveThemeValue
// https://github.com/tailwindlabs/tailwindcss/blob/master/src/util/transformThemeValue.js
const expandAST = ['addBase', 'addComponents', 'addUtilities', 'theme', 'addVariant', 'config', 'corePlugins', 'e', 'matchComponents', 'matchUtilities', 'matchVariant'].map(
  (x) => {
    // key : value
    return t.objectProperty(t.identifier(x), t.identifier(x))
  }
)

function expandAPI() {
  return expandAST
}

export function unwrapThemeFunctionArg(str: string) {
  if (str) {
    const start = str[0]
    const last = str.at(-1)
    if ((start === '"' && last === '"') || (start === "'" && last === "'")) {
      return str.slice(1, -1)
    }
  }
  return str
}

function makeObjectProperty(key: string, value: t.Expression | t.PatternLike) {
  return t.objectProperty(t.stringLiteral(key), value)
}

function addSuffix(value: string, important: boolean) {
  if (typeof value === 'string') {
    return value + (important ? ' !important' : '')
  }
  return value
}

function makeObjectExpression(nodes: Node[]): t.ObjectProperty[] {
  // console.log(nodes)
  // @apply
  // theme('colors.white');

  // theme: <TDefaultValue = Config['theme']>(
  //   path?: string,
  //   defaultValue?: TDefaultValue
  // ) => TDefaultValue

  // css
  // https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/evaluateTailwindFunctions.js

  // plugin
  // https://github.com/tailwindlabs/tailwindcss/blob/master/src/util/createPlugin.js
  // https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/setupContextUtils.js#L298
  // resolvePlugins
  // const rawObj = nodes.reduce<Record<string, Record<string, any>>>((acc, _cur) => {
  //   if (_cur.type === 'rule') {
  //     const cur = _cur as Rule
  //     const selector = cur.selector
  //     if (acc[selector]) {
  //       cur.nodes.reduce()
  //     } else {
  //       acc[selector] = {}
  //     }
  //   }
  //   return acc
  // }, {})

  const res = nodes.reduce<Record<string, t.ObjectProperty[]>>((acc, _cur) => {
    if (_cur.type === 'rule') {
      const cur = _cur as Rule
      const key = cur.selector
      const value = cur.nodes.reduce<Record<string, t.BinaryExpression | t.StringLiteral | t.TemplateLiteral | t.CallExpression>>((cacc, ccur) => {
        if (ccur.type === 'decl') {
          const v = ccur.value
          const important = ccur.important
          const prop = ccur.prop
          // https://tailwindcss.com/docs/adding-custom-styles#adding-component-classes
          // https://tailwindcss.com/docs/functions-and-directives#theme
          const arr = [...v.matchAll(/theme\(([^()]+?)\)/g)]
          if (arr.length > 0) {
            // if (arr.length === 1) {
            const first = arr[0]
            if (first[0] === v) {
              const ccc = t.callExpression(t.identifier('theme'), [t.stringLiteral(unwrapThemeFunctionArg(first[1]))])
              cacc[prop] = important ? t.binaryExpression('+', ccc, t.stringLiteral(' !important')) : ccc
            } else {
              let p = 0
              const quasis: t.TemplateElement[] = []
              const expressions: (t.Expression | t.TSType)[] = []
              // const astP = t.templateLiteral([t.templateElement({ raw: 'calc(100vh - ' })], [t.callExpression(t.identifier('theme'), [t.stringLiteral('spacing.12')])])
              for (let i = 0; i < arr.length; i++) {
                const hit = arr[i]
                if (typeof hit.index === 'number') {
                  const raw = v.slice(p, hit.index)
                  p = hit.index + hit[0].length
                  // const scope = v.slice(hit.index, p)
                  quasis.push(t.templateElement({ raw }))
                  expressions.push(t.callExpression(t.identifier('theme'), [t.stringLiteral(unwrapThemeFunctionArg(hit[1]))]))
                  if (i === arr.length - 1) {
                    quasis.push(t.templateElement({ raw: addSuffix(v.slice(p), important) }))
                  }
                }
              }
              cacc[prop] = t.templateLiteral(quasis, expressions)
            }
          } else {
            cacc[prop] = t.stringLiteral(addSuffix(v, important))
          }
        }
        return cacc //
      }, {})

      // merge the same selector
      if (acc[key]) {
        acc[key].push(
          ...Object.entries(value).map(([prop, value]) => {
            return makeObjectProperty(prop, value)
          })
        )
      } else {
        acc[key] = Object.entries(value).map(([prop, value]) => {
          return makeObjectProperty(prop, value)
        })
      }
    }

    return acc
  }, {})

  return Object.entries(res).map(([key, value]) => {
    return t.objectProperty(
      t.stringLiteral(key),
      t.objectExpression(
        Object.values(
          value.reduce<Record<string, t.ObjectProperty>>((acc, cur) => {
            if (cur.key.type === 'StringLiteral') {
              acc[cur.key.value] = cur
            }
            return acc
          }, {})
        )
      )
    )
  })
}

function getVarName(layer: LayerEnumType) {
  return `_${layer}Css`
}

const pluginNameMap: Record<LayerEnumType, string> = {
  base: 'addBase',
  components: 'addComponents',
  utilities: 'addUtilities'
}

function getFnName(key: LayerEnumType) {
  return pluginNameMap[key]
}

function getLegacyStatement(key: LayerEnumType, ctx: BaseContext) {
  return t.expressionStatement(t.callExpression(t.identifier(getFnName(key)), [t.objectExpression(makeObjectExpression(ctx.getNodes(key)))]))
}

const returnSelfArrowFnAst = t.arrowFunctionExpression([t.identifier('x')], t.identifier('x'))

const declReturnSelfArrowFnName = 'returnSelfNoop'

const declReturnSelfArrowFnAst = t.variableDeclaration('const', [t.variableDeclarator(t.identifier(declReturnSelfArrowFnName), returnSelfArrowFnAst)])

export function createGenerator() {
  // note equal 'plugin'
  const callFnId = '_plugin'
  const pluginName = 'css2TwPlugin'
  const withOptionsParamsId = '_options'
  const traverseMethodName = 'withOptionsWalkCSSRuleObject'
  const requireStatement = t.variableDeclaration('const', [
    t.variableDeclarator(t.identifier(callFnId), t.callExpression(t.identifier('require'), [t.stringLiteral('tailwindcss/plugin')]))
  ])
  const exportsStatement = t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('module'), t.identifier('exports')), t.identifier(pluginName)))

  function getLatestStatement(key: LayerEnumType, ctx: BaseContext) {
    const varName = getVarName(key)
    const fnName = getFnName(key)
    return [
      t.variableDeclaration('const', [
        t.variableDeclarator(
          t.identifier(varName),
          t.callExpression(t.identifier(traverseMethodName), [t.objectExpression(makeObjectExpression(ctx.getNodes(key))), t.stringLiteral(key)])
        )
      ]),
      t.expressionStatement(t.callExpression(t.identifier(fnName), [t.identifier(varName)]))
    ]
  }
  function generate(ctx: BaseContext, opt?: GeneratorOptions) {
    function getPluginBody() {
      const statements = ctx.options?.withOptions
        ? layerNodesKeys.flatMap((layer) => {
            return getLatestStatement(layer, ctx)
          })
        : layerNodesKeys.map((layer) => {
            return getLegacyStatement(layer, ctx)
          })

      const innerPluginAst = t.functionExpression(
        null,
        [t.objectPattern(expandAPI())],

        t.blockStatement(statements)
      )
      if (ctx.options?.withOptions) {
        return t.callExpression(t.memberExpression(t.identifier(callFnId), t.identifier('withOptions')), [
          t.functionExpression(
            null,
            [t.assignmentPattern(t.identifier(withOptionsParamsId), t.objectExpression([]))],
            t.blockStatement([
              // const { withOptionsWalkCSSRuleObject = (x)=>x } = _options
              t.variableDeclaration('const', [
                t.variableDeclarator(
                  t.objectPattern([
                    t.objectProperty(t.identifier(traverseMethodName), t.assignmentPattern(t.identifier(traverseMethodName), t.identifier(declReturnSelfArrowFnName)))
                  ]),
                  t.identifier(withOptionsParamsId)
                )
              ]),
              t.returnStatement(innerPluginAst)
            ])
          ),
          t.functionExpression(null, [t.identifier(withOptionsParamsId)], t.blockStatement([t.returnStatement(t.objectExpression([]))]))
        ])
      }
      return t.callExpression(t.identifier(callFnId), [innerPluginAst])
    }

    const ast = t.file(
      t.program([requireStatement, declReturnSelfArrowFnAst, t.variableDeclaration('const', [t.variableDeclarator(t.identifier(pluginName), getPluginBody())]), exportsStatement])
    )
    const res = babelGenerate(ast, opt)
    return res.code
  }
  return {
    generate
  }
}
