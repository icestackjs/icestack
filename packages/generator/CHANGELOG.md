# @icestack/generator

## 1.0.2

### Patch Changes

- [`ff8f78f`](https://github.com/sonofmagic/icestack/commit/ff8f78f27cf8bf7cbe82b00ea71ae610507991e0) Thanks [@sonofmagic](https://github.com/sonofmagic)! - release 1.3.5

- Updated dependencies [[`ff8f78f`](https://github.com/sonofmagic/icestack/commit/ff8f78f27cf8bf7cbe82b00ea71ae610507991e0)]:
  - @icestack/shared@1.3.5

## 1.0.1

### Patch Changes

- [`14288ba`](https://github.com/sonofmagic/icestack/commit/14288baf04d607a23382e8770042d9a9140d2142) Thanks [@sonofmagic](https://github.com/sonofmagic)! - release cva generate

- Updated dependencies [[`14288ba`](https://github.com/sonofmagic/icestack/commit/14288baf04d607a23382e8770042d9a9140d2142)]:
  - @icestack/shared@1.3.4
