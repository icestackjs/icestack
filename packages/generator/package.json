{
  "name": "@icestack/generator",
  "description": "The generator package of icestack",
  "version": "1.0.3",
  "scripts": {
    "build": "unbuild",
    "test": "vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "sync": "cnpm sync @icestack/generator"
  },
  "keywords": [
    "generate",
    "generator",
    "babel",
    "tailwindcss",
    "unocss",
    "icestack",
    "ui"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    },
    "./*": "./*"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "typesVersions": {
    "*": {
      "*": [
        "./dist/*",
        "./dist/index.d.ts"
      ]
    }
  },
  "dependencies": {
    "@babel/generator": "^7.23.6",
    "@babel/types": "^7.23.9",
    "@icestack/shared": "workspace:^"
  },
  "author": "SonOfMagic <qq1324318532@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/sonofmagic/icestack"
  },
  "homepage": "https://ui.icebreaker.top",
  "repository": {
    "type": "git",
    "url": "https://github.com/sonofmagic/icestack"
  }
}