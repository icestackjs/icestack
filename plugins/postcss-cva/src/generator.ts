import _babelGenerate from '@babel/generator'
import * as t from '@babel/types'
import { defu } from '@icestack/shared'
// import type { VariantProps } from '@icestack/cva'
// https://github.com/babel/babel/issues/15269
function _interopDefaultCompat(e: any) {
  return e && typeof e === 'object' && 'default' in e ? e.default : e
}
export const babelGenerate = _interopDefaultCompat(_babelGenerate) as typeof _babelGenerate

function makeVariants(variants: Record<string, Record<string, string[]>>) {
  const properties = Object.entries(variants).map(([key, value]) => {
    return t.objectProperty(
      t.stringLiteral(key),
      t.objectExpression(
        Object.entries(value).map(([k, v]) => {
          return t.objectProperty(t.stringLiteral(k), t.arrayExpression(v.map(s => t.stringLiteral(s))))
        }),
      ),
    )
  })
  return t.objectExpression(properties)
}

function makeCompoundVariants(compoundVariants: Record<string, string | string[]>[]) {
  return t.arrayExpression(
    compoundVariants.map((x) => {
      const entries = Object.entries(x)
      return t.objectExpression(
        entries.map(([k, v]) => {
          if (!Array.isArray(v)) {
            v = [v]
          }
          return t.objectProperty(
            t.stringLiteral(k),
            t.arrayExpression(
              v.map((x) => {
                return t.stringLiteral(x)
              }),
            ),
          )
        }),
      )
    }),
  )
}

function makeDefaultVariants(defaultVariants: Record<string, string>) {
  return t.objectExpression(
    Object.entries(defaultVariants).map(([key, value]) => {
      return t.objectProperty(t.stringLiteral(key), t.stringLiteral(value))
    }),
  )
}

export function generateCva(options: {
  format?: 'js' | 'ts'
  importFrom?: string
  exports?: Partial<{
    base: boolean
    variants: boolean
    compoundVariants: boolean
    defaultVariants: boolean
  }>
  base: string[]
  variants: Record<string, Record<string, string[]>>
  compoundVariants: Record<string, string | string[]>[]
  defaultVariants: Record<string, string>
}) {
  const baseIdentifier = 'base'
  const variantsIdentifier = 'variants'
  const compoundVariantsIdentifier = 'compoundVariants'
  const defaultVariantsIdentifier = 'defaultVariants'
  const { base, compoundVariants, defaultVariants, format, importFrom, variants, exports } = defu(options, {
    format: 'ts',
    importFrom: 'class-variance-authority',
    base: [],
    variants: {},
    compoundVariants: [],
    defaultVariants: {},
    exports: {
      base: true,
      variants: true,
      compoundVariants: true,
      defaultVariants: true,
    },
  })
  const isTs = format === 'ts'
  const cvaFnName = 'index'
  const body: t.Statement[] = []

  // import
  const specifiers: t.ImportSpecifier[] = [t.importSpecifier(t.identifier('cva'), t.identifier('cva'))]
  if (isTs) {
    const variantPropsNode = t.importSpecifier(t.identifier('VariantProps'), t.identifier('VariantProps'))
    variantPropsNode.importKind = 'type'
    specifiers.push(variantPropsNode)
  }
  // import { cva, type VariantProps } from 'class-variance-authority'
  body.push(t.importDeclaration(specifiers, t.stringLiteral(importFrom)))
  // cva function

  const baseStringArrayExpression = t.arrayExpression(base.map(x => t.stringLiteral(x)))
  const variantsObjectExpression = makeVariants(variants)
  const compoundVariantsArrayExpression = makeCompoundVariants(compoundVariants)
  const defaultVariantsObjectExpression = makeDefaultVariants(defaultVariants)
  if (exports.base) {
    body.push(t.exportNamedDeclaration(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(baseIdentifier), baseStringArrayExpression)])))
  }

  if (exports.variants) {
    body.push(t.exportNamedDeclaration(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(variantsIdentifier), variantsObjectExpression)])))
  }

  if (exports.compoundVariants) {
    body.push(t.exportNamedDeclaration(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(compoundVariantsIdentifier), compoundVariantsArrayExpression)])))
  }

  if (exports.defaultVariants) {
    body.push(t.exportNamedDeclaration(t.variableDeclaration('const', [t.variableDeclarator(t.identifier(defaultVariantsIdentifier), defaultVariantsObjectExpression)])))
  }

  function getConfigParams() {
    const result: t.ObjectProperty[] = []
    result.push(exports.variants ? t.objectProperty(t.identifier('variants'), t.identifier('variants')) : t.objectProperty(t.identifier('variants'), variantsObjectExpression))
    if (exports.compoundVariants) {
      const node = t.objectProperty(t.identifier('compoundVariants'), t.identifier('compoundVariants'))
      isTs && t.addComment(node, 'leading', ' @ts-ignore', true)
      result.push(node)
    }
    else {
      result.push(t.objectProperty(t.identifier('compoundVariants'), compoundVariantsArrayExpression))
    }
    if (exports.defaultVariants) {
      const node = t.objectProperty(t.identifier('defaultVariants'), t.identifier('defaultVariants'))
      isTs && t.addComment(node, 'leading', ' @ts-ignore', true)
      result.push(node)
    }
    else {
      result.push(t.objectProperty(t.identifier('defaultVariants'), defaultVariantsObjectExpression))
    }

    return result
  }
  // cva function
  body.push(
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier(cvaFnName),
        t.callExpression(t.identifier('cva'), [
          // cva CallExpression args
          exports.base ? t.identifier(baseIdentifier) : baseStringArrayExpression,
          t.objectExpression(getConfigParams()),
        ]),
      ),
    ]),
  )
  // cva props
  // export type Props = VariantProps<typeof index>
  if (isTs) {
    body.push(
      t.exportNamedDeclaration(
        t.typeAlias(
          t.identifier('Props'),
          null,
          t.genericTypeAnnotation(t.identifier('VariantProps'), t.typeParameterInstantiation([t.typeofTypeAnnotation(t.genericTypeAnnotation(t.identifier(cvaFnName)))])),
        ),
      ),
    )
  }
  // export default index
  body.push(t.exportDefaultDeclaration(t.identifier(cvaFnName)))
  const ast = t.file(t.program(body))
  return babelGenerate(ast)
}
