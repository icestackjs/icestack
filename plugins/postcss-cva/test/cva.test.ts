import fs from 'node:fs/promises'
import path from 'node:path'
import { generateCva } from '@/generator'
describe('cva', () => {
  it('base', async () => {
    const opt: Parameters<typeof generateCva>[0] = {
      base: ['font-semibold', 'border', 'rounded'],
      variants: {
        intent: {
          primary: ['bg-blue-500', 'text-white', 'border-transparent', 'hover:bg-blue-600'],
          // **or**
          // primary: "bg-blue-500 text-white border-transparent hover:bg-blue-600",
          secondary: ['bg-white', 'text-gray-800', 'border-gray-400', 'hover:bg-gray-100']
        },
        size: {
          small: ['text-sm', 'py-1', 'px-2'],
          medium: ['text-base', 'py-2', 'px-4']
        }
      },
      compoundVariants: [
        {
          intent: ['primary'],
          size: ['medium'],
          class: ['uppercase']
          // **or** if you're a React.js user, `className` may feel more consistent:
          // className: "uppercase"
        }
      ],
      defaultVariants: {
        intent: 'primary',
        size: 'medium'
      }
    }
    opt.exports = {
      base: false,
      compoundVariants: false,
      defaultVariants: false,
      variants: false
    }
    let code = generateCva(opt).code
    expect(code).toMatchSnapshot()
    opt.importFrom = '@icestack/cva'
    await fs.writeFile(path.resolve(__dirname, './fixtures/cva.base.ts'), code)
    opt.format = 'js'
    code = generateCva(opt).code
    expect(code).toMatchSnapshot()
    await fs.writeFile(path.resolve(__dirname, './fixtures/cva.base.js'), code)
    opt.importFrom = 'class-variance-authority'
    opt.format = 'ts'
    code = generateCva(opt).code
    expect(code).toMatchSnapshot()
    await fs.writeFile(path.resolve(__dirname, './fixtures/cva.base.import.ts'), code)
  })
})
