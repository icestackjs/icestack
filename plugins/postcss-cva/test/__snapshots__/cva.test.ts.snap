// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cva > base 1`] = `
"import { cva, type VariantProps } from "class-variance-authority";
const index = cva(["font-semibold", "border", "rounded"], {
  variants: {
    "intent": {
      "primary": ["bg-blue-500", "text-white", "border-transparent", "hover:bg-blue-600"],
      "secondary": ["bg-white", "text-gray-800", "border-gray-400", "hover:bg-gray-100"]
    },
    "size": {
      "small": ["text-sm", "py-1", "px-2"],
      "medium": ["text-base", "py-2", "px-4"]
    }
  },
  compoundVariants: [{
    "intent": ["primary"],
    "size": ["medium"],
    "class": ["uppercase"]
  }],
  defaultVariants: {
    "intent": "primary",
    "size": "medium"
  }
});
export type Props = VariantProps<typeof index>;
export default index;"
`;

exports[`cva > base 2`] = `
"import { cva } from "@icestack/cva";
const index = cva(["font-semibold", "border", "rounded"], {
  variants: {
    "intent": {
      "primary": ["bg-blue-500", "text-white", "border-transparent", "hover:bg-blue-600"],
      "secondary": ["bg-white", "text-gray-800", "border-gray-400", "hover:bg-gray-100"]
    },
    "size": {
      "small": ["text-sm", "py-1", "px-2"],
      "medium": ["text-base", "py-2", "px-4"]
    }
  },
  compoundVariants: [{
    "intent": ["primary"],
    "size": ["medium"],
    "class": ["uppercase"]
  }],
  defaultVariants: {
    "intent": "primary",
    "size": "medium"
  }
});
export default index;"
`;

exports[`cva > base 3`] = `
"import { cva, type VariantProps } from "class-variance-authority";
const index = cva(["font-semibold", "border", "rounded"], {
  variants: {
    "intent": {
      "primary": ["bg-blue-500", "text-white", "border-transparent", "hover:bg-blue-600"],
      "secondary": ["bg-white", "text-gray-800", "border-gray-400", "hover:bg-gray-100"]
    },
    "size": {
      "small": ["text-sm", "py-1", "px-2"],
      "medium": ["text-base", "py-2", "px-4"]
    }
  },
  compoundVariants: [{
    "intent": ["primary"],
    "size": ["medium"],
    "class": ["uppercase"]
  }],
  defaultVariants: {
    "intent": "primary",
    "size": "medium"
  }
});
export type Props = VariantProps<typeof index>;
export default index;"
`;
